### PR TITLE
Camp: Admin and User functionalities

### PR DESCRIPTION
**Pull request by samraarshad63**

### CODE REVIEW

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -29,19 +30,33 @@ def show; end
   def edit; end
 
   def update
-    if @user.update(user_params)
-      redirect_to @user
+    if @user.update(update_user_params)
+      redirect_to admin_users_path
     else
       render 'edit'
+      flash[:alert] = "Couldnt save changes.Try again."
```
Comment (faisalraza) [inline app/controllers/admin/users_controller.rb]: user.errors.full_messages.to_sentence

File: app/controllers/admin/users_controller.rb [hunk 2]
```diff
@@ -29,19 +30,33 @@ def show; end
...
+  end
+
   private
   
   def user_params
     params.require(:user).permit(:keyword)
   end
 
+  def update_user_params
+    if params[:user][:password].blank?
```
Comment (faisalraza) [inline app/controllers/admin/users_controller.rb]: use dig params.dig(:user, :password)

File: app/controllers/admin/users_controller.rb [hunk 3]
```diff
@@ -29,19 +30,33 @@ def show; end
...
+      params.require(:user).permit(:id, :firstname, :lastname, :email, :country, :contact)
+    else
+      params.require(:user).permit(:id, :firstname, :lastname, :email, :password, :country, :contact)
+    end
```
Comment (faisalraza) [inline app/controllers/admin/users_controller.rb]: Update this as discussed in call.

File: app/controllers/users/registrations_controller.rb [hunk 4]
```diff
@@ -1,9 +1,29 @@
 class Users::RegistrationsController < Devise::RegistrationsController
   before_action :update_allowed_parameters
 
+  layout "users", except: [:new]
+
+  def update
+    if @user.update(user_params)
+      redirect_to user_path(current_user)
```
Comment (faisalraza) [inline app/controllers/users/registrations_controller.rb]: move flash above this line.

File: app/controllers/users/registrations_controller.rb [hunk 5]
```diff
@@ -1,9 +1,29 @@
...
+  end
+
   protected
   
   def update_allowed_parameters
     devise_parameter_sanitizer.permit(:sign_up) { |u| u.permit(:firstname, :lastname, :email, :password, :country, :contact, :terms) }
   end
+
+  def user_params
+    if params[:user][:password].blank?
```
Comment (faisalraza) [inline app/controllers/users/registrations_controller.rb]: Update this as discussed in call

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -16,7 +18,30 @@ def update
     end
   end
 
+  def show_camps
+    @camps = Camp.where(status: "active")
```
Comment (faisalraza) [inline app/controllers/users_controller.rb]: Please make a scope of it and use that here i.e scope :active, -> { where(status: "active") }

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -16,7 +18,30 @@ def update
...
+      redirect_to user_personal_info_path
+    else
+      @camp_user = CampUser.new(camp_users_params)
+      @camp_user.user_id = params[:id]
+      @camp_user.save
+      flash[:notice] = "Camp Selected"
+      redirect_to user_personal_path
+    end
+  end
+
```
Comment (faisalraza) [inline app/controllers/users_controller.rb:29]: Please consider updating it like follow: 

def select_camp
  return redirect_to user_personal_info_path, notice: "You have already selected camp" if CampUser.find_by(user_id: params[:id]).present?

  @camp_user = CampUser.new(camp_users_params)
  @camp_user.user_id = params[:id]
  if @camp_user.save
    redirect_to user_personal_path, notice: "Camp Selected"
  else
    redirect_to user_personal_path, alert: @camp_user.errors_.full_messages.to_sentence
  end
end

File: app/models/camp.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    active: "active",
+    inactive: "inactive"
+  }
+
+  enum status: CAMP_STATUS
+
+  def self.search(keyword)
+    return all if keyword.blank?
+
+    where('camp_name LIKE :value', {value: keyword})
```
Comment (faisalraza) [inline app/models/camp.rb]: please use lower in raw query and also downcase the keyword.

File: app/models/camp.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+class Camp < ApplicationRecord
+  has_and_belongs_to_many :locations
+  has_many :users, through: :camp_user
```
Comment (faisalraza) [inline app/models/camp.rb]: has_many :camp_users
has_many :users, through: :camp_users

File: app/models/user.rb [hunk 10]
```diff
@@ -2,12 +2,14 @@ class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :confirmable, :registerable, :recoverable, :rememberable, :validatable
+
+  has_one :camp, through: :camp_user
```
Comment (faisalraza) [inline app/models/user.rb:7]: Please fix and update this association

File: app/views/admin/camps/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  </div>
+
+  <div class="field">
+    <%= f.label :end_date %>
+    <p><%= f.date_field :end_date, required: true %></p>
+  </div>
+
+  <p>
+    <%= collection_check_boxes(:camp, :location_ids, Location.all, :id, :name) do |b| %>
+      <p><%= b.label("data-value": b.value) { b.check_box + " " + b.text } %></p>
```
Comment (faisalraza) [inline app/views/admin/camps/new.html.erb:22]: Can we please string interpolation here?

File: app/views/admin/users/index.csv.erb [hunk 12]
```diff
@@ -1,4 +1,4 @@
-<%- User::CSV_HEADERS -%>
-<%- @users.each do |user| -%>
+<%= CSV.generate_line User::CSV_HEADERS %>
+<%- @all_users.each do |user| -%>
```
Comment (faisalraza) [inline app/views/admin/users/index.csv.erb]: Can we please update it to a better name like csv_users_list ?

File: app/controllers/admin/users_controller.rb [hunk 13]
```diff
@@ -29,10 +30,27 @@ def show; end
...
+        redirect_to admin_users_path, notice: 'User profile updated successfully'
+      else
+        render 'edit'
+      end
+    else
+      if @user.update(update_user_params)
+        redirect_to admin_users_path, notice: 'User profile updated successfully'
+      else
+        render 'edit'
+      end
```
Comment (faisalraza) [inline app/controllers/admin/users_controller.rb]: Can we please reduce the if/elses here.

File: app/controllers/admin/users_controller.rb [hunk 14]
```diff
@@ -29,10 +30,27 @@ def show; end
...
+    end
+  end
+
+  def destroy
+    if @user.destroy
+      redirect_to admin_users_path, notice: 'User destroyed successfully'
     else
-      render 'edit'
+      flash[:alert] = @user.errors.full_messages.to_sentence
+      redirect_to admin_users_path
```
Comment (faisalraza) [inline app/controllers/admin/users_controller.rb]: def destroy
    if @user.destroy
      flash[:notice] =  'User destroyed successfully'
    else
      flash[:alert] = @user.errors.full_messages.to_sentence
    end    
    redirect_to admin_users_path
end

File: app/controllers/users/registrations_controller.rb [hunk 15]
```diff
@@ -1,9 +1,31 @@
...
+        render 'edit'
+      end
+    else
+      if @user.update(user_params)
+        redirect_to user_path(current_user), notice: 'Profile updated successfully'
+      else
+        render 'edit'
+      end
+    end
+  end
```
Comment (faisalraza) [inline app/controllers/users/registrations_controller.rb:17]: Please reduce if/elses here

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -8,15 +8,32 @@ def show; end
...
+  def show_camps
+    @camps = Camp.active
+  end
+
+  def select_camp
+    return redirect_to user_personal_info_path, notice: "You have already selected camp" if CampUser.find_by(user_id: params[:id]).present?
+
+    @camp_user = CampUser.new(camp_users_params)
+    @camp_user.user_id = params[:id]
```
Comment (faisalraza) [inline app/controllers/users_controller.rb]: what if i intentionally send someone else's id by manipulating front end? Would this assign camp to that user?

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -8,15 +8,32 @@ def show; end
...
+    if @camp_user.save
+      redirect_to user_personal_path, notice: 'Camp Selected'
     else
-      render 'edit'
+      flash[:alert] = @user.errors.full_messages.to_sentence
+      redirect_to user_personal_path
```
Comment (faisalraza) [inline app/controllers/users_controller.rb]: This can be refactored by taking out the common things.

File: app/models/camp.rb [hunk 18]
```diff
@@ -0,0 +1,20 @@
+class Camp < ApplicationRecord
+  has_and_belongs_to_many :locations
+  has_many :camp_users
```
Comment (faisalraza) [inline app/models/camp.rb]: Do we need to destroy these when a camp is destroyed?

File: app/models/user.rb [hunk 19]
```diff
@@ -2,21 +2,27 @@ class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :confirmable, :registerable, :recoverable, :rememberable, :validatable
+
+  has_one :camp_user
```
Comment (faisalraza) [inline app/models/user.rb]: do we need to destroy this if this user is destroyed?

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (samraarshad63) [reply app/controllers/users_controller.rb]: yes! i should make use of current_user here instead

File: app/models/user.rb [hunk 20]
```diff
@@ -2,21 +2,27 @@ class User < ApplicationRecord
...
   PER_PAGE = 3
   USER_ROLES = { 
     user: "user"
-  }.freeze
+  }
   CSV_HEADERS = ['Firstname', 'Lastname', 'Email', 'Contact', 'Country']
   
   enum user_role: USER_ROLES
 
+  scope :active, -> { where(status: "active") }
```
Comment (faisalraza) [inline app/models/user.rb]: do we have a status column in user?

Comment (samraarshad63) [reply app/models/user.rb]: no! This should be in the camp model

File: config/routes.rb [hunk 21]
```diff
@@ -4,9 +4,13 @@
 
   root 'home#index', as: :home
   get "/pages/:page" => "pages#show", as: :pages
+  get "/users/:id/camps" => "users#show_camps", as: :user_show_camps
+  post "/users/:id/camp" => "users#select_camp", as: :user_select_camp
+  get "/users/:id/camp/step-1" => "users#personal_info", as: :user_personal_info
```
Comment (faisalraza) [inline config/routes.rb]: Can we define these in some better way?

File: app/views/admin/users/edit.html.erb [hunk 22]
```diff
@@ -3,35 +3,41 @@
...
+    <%= f.label :email, "Email Address" %>
+    <p><%= f.email_field :email, autocomplete: "email" %></p>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
   </div>
 
-  <%= f.submit %>
+  <%= render "layouts/error_messages", object: @user %>
```
Comment (faisalraza) [inline app/views/admin/users/edit.html.erb]: Shouldn't it be after the form_for tag?

File: app/views/devise/sessions/new.html.erb [hunk 23]
```diff
@@ -1,26 +1,30 @@
...
+  <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
 
-  <div class="field">
-    <%= f.label :password %><br />
-    <%= f.password_field :password, autocomplete: "current-password" %>
-  </div>
+    <div class="field">
+      <%= f.label :email %>
+      <p><%= f.email_field :email, autofocus: true, autocomplete: "email" %></p>
+    </div>
```
Comment (faisalraza) [inline app/views/devise/sessions/new.html.erb:9]: Extra Spaces.

File: config/routes.rb [hunk 24]
```diff
@@ -4,11 +4,18 @@
 
   root 'home#index', as: :home
   get "/pages/:page" => "pages#show", as: :pages
-
+
```
Comment (faisalraza) [inline config/routes.rb]: Spacing issue.

**Pull request merged**